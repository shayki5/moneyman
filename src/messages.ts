import {
  TransactionStatuses,
  TransactionTypes,
} from "israeli-bank-scrapers/lib/transactions.js";
import {
  AccountScrapeResult,
  SaveStats,
  Transaction,
  TransactionRow,
} from "./types";

export function getSummaryMessages(results: Array<AccountScrapeResult>) {
  const accountsSummary = results.flatMap(({ result, companyId }) => {
    if (!result.success) {
      return `\t❌ [${companyId}] ${result.errorType}${
        result.errorMessage ? `\n\t\t${result.errorMessage}` : ""
      }`;
    }
    return result.accounts?.map(
      (account) =>
        `\t✔️ [${companyId}] ${account.accountNumber}: ${account.txns.length}`,
    );
  });

  const { pending, completed } = transactionsByStatus(results);

  return `
${transactionsString(pending, completed)}

Accounts updated:
${accountsSummary.join("\n") || "\t😶 None"}

Pending txns:
${transactionList(pending) || "\t😶 None"}
`.trim();
}

function transactionsString(
  pending: Array<Transaction>,
  completed: Array<Transaction>,
) {
  const total = pending.length + completed.length;

  return `
${total} transactions scraped.
${
  total > 0 ? `(${pending.length} pending, ${completed.length} completed)` : ""
}`.trim();
}

function transactionAmount(t: Transaction): number {
  switch (t.type) {
    case TransactionTypes.Normal:
      switch (t.status) {
        case TransactionStatuses.Pending:
          return t.originalAmount;
        case TransactionStatuses.Completed:
          return t.chargedAmount;
      }
    case TransactionTypes.Installments:
      return t.chargedAmount;
  }
}

function transactionString(t: Transaction) {
  const amount = transactionAmount(t);

  const sign = amount < 0 ? "-" : "+";
  const absAmount = Math.abs(amount).toFixed(2);

  return `${t?.description}:\t${sign}${absAmount}${
    t.originalCurrency === "ILS" ? "" : ` ${t.originalCurrency}`
  }`;
}

function transactionList(transactions: Array<Transaction>, indent = "\t") {
  return transactions.map((t) => `${indent}${transactionString(t)}`).join("\n");
}
export function saving(storage: string) {
  return `📝 ${storage} Saving...`;
}

export function saved(stats: SaveStats): string {
  return `
📝 ${stats.name} (${stats.table})
\t${stats.added} added
\t${stats.skipped} skipped (${stats.existing} existing, ${
    stats.pending
  } pending)
${highlightedTransactionsString(stats.highlightedTransactions, 1)}`.trim();
}

function transactionsByStatus(results: Array<AccountScrapeResult>) {
  const allTxns = results
    .flatMap(({ result }) =>
      result.accounts?.flatMap((account) => account?.txns),
    )
    .filter((t): t is Transaction => t !== undefined);

  const pendingTxns = allTxns.filter(
    (t) => t.status === TransactionStatuses.Pending,
  );
  const scrapedTxns = allTxns.filter(
    (t) => t.status === TransactionStatuses.Completed,
  );

  return {
    pending: pendingTxns,
    completed: scrapedTxns,
  };
}

function highlightedTransactionsString(
  groups: Record<string, TransactionRow[]> | undefined,
  indent = 0,
) {
  if (!groups || Object.keys(groups).length === 0) {
    return "";
  }

  const indentString = "\t".repeat(indent);
  const groupsString = Object.entries(groups)
    .filter(([_, txns]) => txns.length > 0)
    .map(([name, txns]) => {
      const transactionsString = transactionList(txns, `${indentString}\t`);
      return `${indentString}${name}:\n${transactionsString}`;
    });

  if (groupsString.length === 0) {
    return "";
  }

  return `${indentString}${"-".repeat(5)}\n${groupsString}`;
}
