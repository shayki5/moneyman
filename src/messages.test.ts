import { CompanyTypes } from "israeli-bank-scrapers";
import { getSummaryMessages, saved } from "./messages";
import {
  AccountScrapeResult,
  SaveStats,
  Transaction,
  TransactionRow,
} from "./types";
import {
  TransactionStatuses,
  TransactionTypes,
} from "israeli-bank-scrapers/lib/transactions";
import { ScraperErrorTypes } from "israeli-bank-scrapers/lib/scrapers/errors";

describe("messages", () => {
  describe("getSummaryMessages", () => {
    it("should return a summary message", () => {
      const results: Array<AccountScrapeResult> = [
        {
          companyId: CompanyTypes.max,
          result: {
            success: true,
            accounts: [
              {
                accountNumber: "account1",
                txns: [
                  transaction({
                    chargedAmount: -20,
                    originalAmount: -100,
                    description: "ILS",
                    chargedCurrency: "ILS",
                    originalCurrency: "USD",
                  }),
                ],
              },
              {
                accountNumber: "account2",
                txns: [
                  transaction({
                    type: TransactionTypes.Installments,
                  }),
                  transaction({
                    status: TransactionStatuses.Pending,
                    originalAmount: 20,
                  }),
                  transaction({
                    status: TransactionStatuses.Pending,
                    originalAmount: 20,
                    originalCurrency: "USD",
                  }),
                  transaction({
                    status: TransactionStatuses.Pending,
                    originalAmount: -20,
                  }),
                  transaction({
                    description: "description2",
                    memo: "memo2",
                  }),
                  transaction({
                    chargedAmount: 20,
                    originalAmount: 5,
                    originalCurrency: "USD",
                  }),
                ],
              },
            ],
          },
        },
      ];

      const stats: Array<SaveStats> = [
        {
          name: "Storage 1",
          table: "TheTable",
          total: 1,
          added: 2,
          pending: 3,
          skipped: 4,
          existing: 5,
        },
        {
          name: "Storage 2",
          table: "TheTable",
          total: 6,
          added: 7,
          pending: 8,
          skipped: 9,
          existing: 10,
          highlightedTransactions: {
            Group1: [
              {
                account: "account1",
                companyId: CompanyTypes.max,
                hash: "hash1",
                type: TransactionTypes.Normal,
                date: new Date().toISOString(),
                processedDate: new Date().toISOString(),
                description: "description1",
                originalAmount: 10,
                uniqueId: "uniqueId1",
                originalCurrency: "ILS",
                chargedAmount: 10,
                status: TransactionStatuses.Completed,
              },
            ],
          },
        },
      ];

      const summary = getSummaryMessages(results);
      expect(summary).toMatchSnapshot();

      const saveSummaries = stats.map(saved);
      expect(saveSummaries).toMatchSnapshot();
    });

    it("should return a summary message with no results", () => {
      const results: Array<AccountScrapeResult> = [];

      const stats: Array<SaveStats> = [];

      const summary = getSummaryMessages(results);
      expect(summary).toMatchSnapshot();

      const saveSummaries = stats.map(saved);
      expect(saveSummaries).toMatchSnapshot();
    });

    it("should return a summary message with failed results", () => {
      const results: Array<AccountScrapeResult> = [
        {
          companyId: CompanyTypes.max,
          result: {
            success: false,
            errorType: ScraperErrorTypes.Generic,
            errorMessage: "Some error message",
          },
        },
        {
          companyId: CompanyTypes.hapoalim,
          result: {
            success: false,
            errorType: ScraperErrorTypes.ChangePassword,
          },
        },
        {
          companyId: CompanyTypes.hapoalim,
          result: {
            success: true,
            accounts: [
              {
                accountNumber: "account1",
                txns: [transaction({})],
              },
            ],
          },
        },
      ];

      const stats: Array<SaveStats> = [];

      const summary = getSummaryMessages(results);
      expect(summary).toMatchSnapshot();

      const saveSummaries = stats.map(saved);
      expect(saveSummaries).toMatchSnapshot();
    });

    it("should return a summary message with installments", () => {
      const transactions = [
        transaction({
          type: TransactionTypes.Installments,
          chargedAmount: 20,
          originalAmount: 100,
          description: "should be +20",
        }),
        transaction({
          type: TransactionTypes.Installments,
          chargedAmount: -20,
          originalAmount: -100,
          description: "should be -20",
        }),
      ];

      const results: Array<AccountScrapeResult> = [
        {
          companyId: CompanyTypes.max,
          result: {
            success: true,
            accounts: [
              {
                accountNumber: "account1",
                txns: transactions,
              },
            ],
          },
        },
      ];

      const stats: Array<SaveStats> = [
        {
          name: "Storage 1",
          table: "TheTable",
          total: 1,
          added: 2,
          pending: 3,
          skipped: 4,
          existing: 5,
          highlightedTransactions: {
            SomeGroup: transactions.map<TransactionRow>((t) => ({
              account: "account1",
              companyId: CompanyTypes.max,
              hash: "hash1",
              uniqueId: "uniqueId1",
              ...t,
            })),
          },
        },
      ];

      const summary = getSummaryMessages(results);
      expect(summary).toMatchSnapshot();

      const saveSummaries = stats.map(saved);
      expect(saveSummaries).toMatchSnapshot();
    });

    it("should not add empty groups", () => {
      const stats: Array<SaveStats> = [
        {
          name: "Storage",
          table: "TheTable",
          total: 1,
          added: 0,
          pending: 0,
          skipped: 0,
          existing: 0,
          highlightedTransactions: {
            Group1: [],
            Group2: [
              {
                ...transaction({}),
                hash: "hash1",
                uniqueId: "uniqueId1",
                account: "account1",
                companyId: CompanyTypes.max,
              },
            ],
          },
        },
      ];

      const saveSummaries = stats.map(saved);
      expect(saveSummaries).toMatchSnapshot();
    });

    it("should not add empty groups", () => {
      const stats: Array<SaveStats> = [
        {
          name: "Storage",
          table: "TheTable",
          total: 1,
          added: 0,
          pending: 0,
          skipped: 0,
          existing: 0,
          highlightedTransactions: {
            Group1: [],
          },
        },
      ];

      const saveSummaries = stats.map(saved);
      expect(saveSummaries).toMatchSnapshot();
    });
  });
});

function transaction(t: Partial<Transaction>): Transaction {
  return {
    type: TransactionTypes.Normal,
    date: new Date().toISOString(),
    processedDate: new Date().toISOString(),
    description: "description1",
    originalAmount: 10,
    originalCurrency: "ILS",
    chargedAmount: t.status === TransactionStatuses.Pending ? 0 : 10,
    status: TransactionStatuses.Completed,
    ...t,
  };
}
