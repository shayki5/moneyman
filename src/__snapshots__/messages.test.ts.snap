// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`messages getSummaryMessages should not add empty groups 1`] = `
[
  "📝 Storage (TheTable)
	0 added
	0 skipped (0 existing, 0 pending)
	-----
	Group2:
		description1:	+10.00",
]
`;

exports[`messages getSummaryMessages should not add empty groups 2`] = `
[
  "📝 Storage (TheTable)
	0 added
	0 skipped (0 existing, 0 pending)",
]
`;

exports[`messages getSummaryMessages should return a summary message 1`] = `
"7 transactions scraped.
(3 pending, 4 completed)

Accounts updated:
	✔️ [max] account1: 1
	✔️ [max] account2: 6

Pending txns:
	description1:	+20.00
	description1:	+20.00 USD
	description1:	-20.00"
`;

exports[`messages getSummaryMessages should return a summary message 2`] = `
[
  "📝 Storage 1 (TheTable)
	2 added
	4 skipped (5 existing, 3 pending)",
  "📝 Storage 2 (TheTable)
	7 added
	9 skipped (10 existing, 8 pending)
	-----
	Group1:
		description1:	+10.00",
]
`;

exports[`messages getSummaryMessages should return a summary message with failed results 1`] = `
"1 transactions scraped.
(0 pending, 1 completed)

Accounts updated:
	❌ [max] GENERIC
		Some error message
	❌ [hapoalim] CHANGE_PASSWORD
	✔️ [hapoalim] account1: 1

Pending txns:
	😶 None"
`;

exports[`messages getSummaryMessages should return a summary message with failed results 2`] = `[]`;

exports[`messages getSummaryMessages should return a summary message with installments 1`] = `
"2 transactions scraped.
(0 pending, 2 completed)

Accounts updated:
	✔️ [max] account1: 2

Pending txns:
	😶 None"
`;

exports[`messages getSummaryMessages should return a summary message with installments 2`] = `
[
  "📝 Storage 1 (TheTable)
	2 added
	4 skipped (5 existing, 3 pending)
	-----
	SomeGroup:
		should be +20:	+20.00
		should be -20:	-20.00",
]
`;

exports[`messages getSummaryMessages should return a summary message with no results 1`] = `
"0 transactions scraped.

Accounts updated:
	😶 None

Pending txns:
	😶 None"
`;

exports[`messages getSummaryMessages should return a summary message with no results 2`] = `[]`;
